### PR TITLE
[SPARK-22591][SQL] GenerateOrdering shouldn't change CodegenContext.INPUT_ROW

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

When I played with codegen in developing another PR, I found the value of `CodegenContext.INPUT_ROW` is not reliable. Under wholestage codegen, it is assigned to null first and then suddenly changed to `i`. 

The reason is `GenerateOrdering` changes `CodegenContext.INPUT_ROW` but doesn't restore it back.

## How was this patch tested?

Added test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @kiszk

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala [hunk 1]
```diff
@@ -156,4 +156,13 @@ class OrderingSuite extends SparkFunSuite with ExpressionEvalHelper {
       assert(genOrdering.compare(rowB1, rowB2) < 0)
     }
   }
+
+  test("SPARK-22591: GenerateOrdering shouldn't change ctx.INPUT_ROW") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: does it cause any bugs?

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: I'm working to support wholestage codegen when generating expression codes safe from 64k limit. When there is not `INPUT_ROW` in context but we wrongly set a `INPUT_ROW` value, a non-existing `InternalRow i` will be added into function parameters.

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: As we use `INPUT_ROW` a lot in codegen framework, it is risky to change its value without restoring it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala [hunk 2]
```diff
@@ -72,6 +72,7 @@ object GenerateOrdering extends CodeGenerator[Seq[SortOrder], Ordering[InternalR
    * Generates the code for ordering based on the given order.
    */
   def genComparisons(ctx: CodegenContext, ordering: Seq[SortOrder]): String = {
+    val oldInputRow = ctx.INPUT_ROW
     val comparisons = ordering.map { order =>
       val oldCurrentVars = ctx.currentVars
```
[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: Yes.

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: how about
```
val oldInputRow = ...
val oldCurrentVars = ...
val inputRow = "i"
ctx.INPUT_ROW = inputRow
ctx.currentVars = null
val comparisons = ...
ctx.INPUT_ROW = oldInputRow
ctx.currentVars = oldCurrentVars
...
```

[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: Ok. Looks good.

**Comment by viirya** [comment]
This test `org.apache.spark.sql.catalyst.expressions.CastSuite.SPARK-22500: cast for struct should not generate codes beyond 64KB` seems flaky? I saw it fails with `java.lang.OutOfMemoryError: GC overhead limit exceeded` in other PR too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala [hunk 3]
```diff
@@ -149,10 +151,14 @@ object GenerateOrdering extends CodeGenerator[Seq[SortOrder], Ordering[InternalR
...
     // returns an inlined block
-    s"""
+    val finalCode = s"""
        |InternalRow ${ctx.INPUT_ROW} = null;
        |$code
      """.stripMargin
+    // Restore original currentVars and INPUT_ROW.
+    ctx.currentVars = oldCurrentVars
+    ctx.INPUT_ROW = oldInputRow
+    finalCode
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]:
```
ctx.currentVars = oldCurrentVars
ctx.INPUT_ROW = oldInputRow
s"""
  |InternalRow $inputRow = null;
  |$code
 """.stripMargin
```

**Comment by cloud-fan** [comment]
I created https://issues.apache.org/jira/browse/SPARK-22595 to track the flaky test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala]: Yes, thanks.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2!

**Pull request closed without merging**